pylint/checkers/base/name_checker/checker.py

```diff
@@ -40,7 +40,8 @@
 DEFAULT_PATTERNS = {
     "typevar": re.compile(
         r"^_{0,2}(?:[^\W\da-z_]+|(?:[^\W\da-z_]+[^\WA-Z_]+)+T?(?<!Type))(?:_co(?:ntra)?)?$"
-    )
+    ),
+    "typealias": re.compile(r"^_{0,2}[A-Z]+[a-z]+([A-Z][a-z]+)*$"),
```

Comment:
```suggestion
    "typealias": re.compile(r"^_{0,2}(?!T[A-Z]|Type)[A-Z]+[a-z]+(?:[A-Z][a-z]+)*$"),
```
Would this be a solution?
The negative lookahead would prevent cases like `TBadName` and `TypeTodo` but still allow `TodoType`. `TodoType` and `TypeTodo` could also be added as test cases.

Reply:
Yeah, although it would still disallow `TTCircuit` as in the circuit of a [tt race](https://en.wikipedia.org/wiki/Isle_of_Man_TT).

That's really far fetched but I don't really like how this is such an arbitrary rule that is really difficult to figure out if you stumble upon it. I don't think `TTCircuit` will be used very often but there might be real uses of `TT` at the start and I think it would be very frustrating to figure out why `pylint` disallows them.

What would you think about using the current pattern as the default value and using your suggestion as our own overwrite? We could potentially even add our own overwrite in the documentation of the message?

Reply:
Getting the regex 100% correct is pretty difficult. What should we do for the pandas cases then (`DatetimeNaTType`, `IntervalOrNA`, ...)?

IMO we should try to find something that works for most cases and can prevent common issues. The remaining ones should be handled with `pylint: disable[invalid-name]`. In this case I believe it's a much more prevalent error to use a `T` or `Type` prefix then something like `TTCircuit`, so I think it would be worth it.

I just opened #7322 to fix the same issue for the `TypeVar` name style as I've seen `_TRemote` and `_TCommand` in HomeAssistant lately.

Reply:
I agree that `T` and `Type` will be much more prevalent but I feel like flagging a `TTCircuit` would be considered a false positive. That's why I think we might want to be a little less strict by default and let users be more strict themselves. For me it falls within the false negatives vs. false positives preference, where I think the former is always more preferable.

Reply:
> For me it falls within the false negatives vs. false positives preference, where I think the former is always more preferable.

Usually I agree here. However, I would consider `invalid-name` to be at least somewhat special. It's one of the few checks which creates false-positives in order to highlight certain errors and I would think users might already be aware of that. Just an example, in HomeAssistant about 7-8% of all `pylint: disable` comments are for `invalid-name` (53 of 697).

Reply:
Hmm doesn't that indicate that `pylint` is being too strict then? The main issue I have is that it is pretty hard to know why `TTCircuit` failed based on the `doesn't conform to predefined naming style` message. If we can find a good way to communicate that T's aren't allowed without having to open the pylint documentation I would be more than fine with disallowing them. We have some other issues about `invalid-name` and it being unclear why stuff is being flagged as well, I'd like to avoid those.

Reply:
Also, how would that work with the new design proposed in #7305 ? Explaining better why invalid-name was triggered is a big priority.

Reply:
I don't think it's that big of deal in practice. At least in my experience the `pylint: disable=invalid-name` is added quickly and everybody is happy.

I think of this check more like a pointer
```
Hey here could be an error, please check that.
But if that's what you want, it's also fine.
```

Maybe we need a new error category for formatting / style issues 🤔

Reply:
(I was thinking of #2018  is the second most up-voted issues in the repo, but the regex is hiding the fact that the name is too short, #7305 aims to separate the two. Explaining *everything* in the regex is lower priority indeed)

Reply:
> Hey here could be an error, please check that. But if that's what you want, it's also fine.

Isn't that "convention" ? It's ok to not follow convention.

Reply:
> > Hey here could be an error, please check that. But if that's what you want, it's also fine.
> 
> Isn't that "convention" ? It's ok to not follow convention.

True 😄
In that case, we might only need to better explain that it's just a convention and nothing more.

Reply:
The issue might be that pep8 is wildly adopted (to a point that a lot of junior dev allow more time for PEP8 than to writing code that actually bring value). But PEP8 does not say anything about typevar/typealias yet so we're mixing a very popular convention with one that is less established.

Reply:
I have updated to use @cdce8p's suggestion and hope that the additional comments in `details.rst` will prevent any issues. I think Marc is right that we should try to enforce good behaviour here rather than prevent 2/3 opened issues about this.